[FLINK-4246] Allow Specifying Multiple Metrics Reporters

aljoscha opened the pull request:
This also updates documentation and tests.

Reporters can now be specified like this:

metrics.reporters: foo,bar

metrics.reporter.foo.class: JMXReporter.class
metrics.reporter.foo.port: 10

metrics.reporter.bar.class: GangliaReporter.class
metrics.reporter.bar.port: 11
metrics.reporter.bar.interval: 10 SECONDS
metrics.reporter.bar.something: 42

R: @StephanEwen and @zentol for review

zentol commented on docs/apis/metrics.md:
```diff
@@ -227,14 +227,25 @@ or by assigning unique names to jobs and operators.
+- `metrics.reporter.<name>.class`: The reporter class to use for the reporter named `<name>`.
+- `metrics.reporter.<name>.interval`: The reporter interval to use for the reporter named `<name>`.
+
+All reporters must at least have the `class` config, some allow specifying a reporting `interval`. Bellow,
```
type: Bellow -> Below

zentol replied:
i would `class` **property** instead of config.

zentol commented on flink-core/src/main/java/org/apache/flink/metrics/MetricRegistry.java:
```diff
@@ -75,77 +78,92 @@ public MetricRegistry(Configuration config) {
+
+				final String className = reporterConfig.getString(ConfigConstants.METRICS_REPORTER_CLASS, null);
+				if (className == null) {
+					throw new IllegalStateException("No reporter class set for reporter " + namedReporter);
```
The MetricRegistry constructor should not throw any exceptions.

zentol commented on flink-core/src/main/java/org/apache/flink/metrics/MetricRegistry.java:
```diff
@@ -75,77 +78,92 @@ public MetricRegistry(Configuration config) {
+					reporterInstance.open(reporterConfig);
+
+					if (reporterInstance instanceof Scheduled) {
+						ensureExecutorRunning();
```
do we really need this method? it is only used once.

zentol commented on flink-core/src/main/java/org/apache/flink/metrics/MetricRegistry.java:
```diff
@@ -198,8 +220,13 @@ public void register(Metric metric, String metricName, AbstractMetricGroup group
-				reporter.notifyOfRemovedMetric(metric, metricName, group);
+			if (reporters != null) {
+				for (MetricReporter reporter : reporters) {
+
```
let's remove this empty line so the code is formatting-wise identical to `register()`

zentol commented on flink-core/src/main/java/org/apache/flink/metrics/MetricRegistry.java:
```diff
@@ -263,6 +276,7 @@ private ReporterTask(Scheduled reporter) {
 
 		@Override
 		public void run() {
+			System.out.println("TRYING REPORT");
```
leftover `println()`

zentol commented on docs/apis/metrics.md:
```diff
@@ -227,14 +227,25 @@ or by assigning unique names to jobs and operators.
-  - Example: --host localhost --port 9010
-- `metrics.reporter.interval`: The interval between reports.
-  - Example: 10 SECONDS
+Metrics can be exposed to an external system by configuring a one or several reporters in `conf/flink-conf.yaml`.
```
typo: configuring a one -> configuring one

zentol commented on docs/apis/metrics.md:
````diff
@@ -227,14 +227,25 @@ or by assigning unique names to jobs and operators.
+Example reporter configuration for using the built-in JMX reporter:
+
+```
+metrics.reporters: my_jmx_reporter
````
i think we should show the configuration for multiple reporters

zentol commented on docs/apis/metrics.md:
```diff
@@ -227,14 +227,25 @@ or by assigning unique names to jobs and operators.
-  - Example: 10 SECONDS
+Metrics can be exposed to an external system by configuring a one or several reporters in `conf/flink-conf.yaml`.
+
+- `metrics.reporters`: The list of named reporters, i.e. "foo,bar".
```
since we have an example configuration just below we can omit the example here.

zentol commented on docs/apis/metrics.md:
```diff
@@ -227,14 +227,25 @@ or by assigning unique names to jobs and operators.
+metrics.reporters: my_jmx_reporter
+
+metrics.reporter.my_jmx_reporter.class: org.apache.flink.metrics.reporter.JMXReporter
+metrics.reporter.my_jmx_reporter.port: 9020-940
```
i think you're missing a zero; 940 -> 9040

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -651,14 +651,38 @@
 
 	// ---------------------------- Metrics -----------------------------------
 
+	// Per reporter:
```
i think this comment and the "For all reporters" don't stand out enough to be included. My first impression was that these were left-overs.

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -651,14 +651,38 @@
+	 * <pre>{@code
+	 * metrics.reporters = foo, bar
+	 *
+	 * metrics.reporter.foo.class = FooReporter.class
```
shouldn't the classes be fully qualified?

zentol commented on flink-core/src/main/java/org/apache/flink/configuration/ConfigConstants.java:
```diff
@@ -651,14 +651,38 @@
-	
-	/** A list of named parameters that are passed to the reporter. */
-	public static final String METRICS_REPORTER_ARGUMENTS = "metrics.reporter.arguments";
+	public static final String METRICS_REPORTER_CLASS = "class";
```
I think we should label both class and interval as suffix keys.

zentol commented on flink-core/src/main/java/org/apache/flink/metrics/MetricRegistry.java:
```diff
@@ -75,77 +78,92 @@ public MetricRegistry(Configuration config) {
+		} else {
+			// we have some reporters so
+			String[] namedReporters = definedReporters.split(",");
+			for (String namedReporter : namedReporters) {
```
Can it happen that namedReporters is an empty String?

aljoscha replied:
In that case the array should be empty and we should not enter the loop, right?

zentol replied:
yes. but this will go completely unnoticed to a user since nothing is being logged.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/util/DelegatingConfigurationTest.java:
```diff
@@ -28,7 +28,7 @@
 import java.util.Comparator;
 
```
this test should be removed, as it was moved into flink-core.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/JobManagerMetricTest.java:
```diff
@@ -51,9 +51,11 @@ public void testJobManagerMetricAccess() throws Exception {
 
-		flinkConfiguration.setString(ConfigConstants.METRICS_REPORTER_CLASS, "org.apache.flink.metrics.reporter.JMXReporter");
+		flinkConfiguration.setString(ConfigConstants.METRICS_REPORTERS_LIST, "test");
+		flinkConfiguration.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test.class", "org.apache.flink.metrics.reporter.JMXReporter");
```
shouldn't we uses the ConfigConstants.METRICS_REPORTER_CLASS here instead of `.class` ? I know it blows up the line, but it still is a defined constant. (the same applies to ExecutionGraphMetricsTest)

zentol commented on flink-metrics/flink-metrics-statsd/src/test/java/org/apache/flink/metrics/statsd/StatsDReporterTest.java:
```diff
@@ -135,11 +138,14 @@ public void testStatsDHistogramReporting() throws Exception {
 			receiverThread.start();
 
 			int port = receiver.getPort();
+			System.out.println("PORT: " + port);
```
leftover `println()` ?

zentol commented on flink-core/src/test/java/org/apache/flink/metrics/reporter/JMXReporterTest.java:
```diff
@@ -129,17 +134,25 @@ public Integer getValue() {
+		cfg.setString(ConfigConstants.METRICS_REPORTERS_LIST, "test1,test2");
+
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test1.class", JMXReporter.class.getName());
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test1.port", "9020-9035");
```
why are you using a different port range?

zentol commented:
The usage of the `DelagatingConfiguration` poses a few problems in regards to #2226. After that PR reporters will no longer work on a `Configuration` but instead a `MetricConfig` object that extends `Properties`. For this to work certain key have to be extracted from the `Configuration` and passes into the `MetricConfig`. Naturally, for this to work you must know the names of all keys required. This doesn't work for arbitrary user-defined settings though. We have to revert back to a `metrics.reporter.<name>.arguments` property.

aljoscha commented:
Thanks for the thorough review @zentol! I'm addressing the comments in a new commit.

About the move to `MetricConfig`, I think the Configuration could be extended for that case to move all parameters with a certain prefix.

zentol commented:
That would imply iterating over the entire configuration and checking for the prefix, correct?

zentol commented on docs/apis/metrics.md:
```diff
@@ -229,21 +229,25 @@ or by assigning unique names to jobs and operators.
-metrics.reporter.my_jmx_reporter.port: 9020-940
+metrics.reporter.my_jmx_reporter.port: 9020-9040
+
+metrics.reporter.my_jmx_reporter.class: org.apache.flink.metrics.graphite.GraphiteReporter
```
should be named my_other_reporter

zentol commented on flink-examples/flink-examples-streaming/src/main/java/org/apache/flink/streaming/examples/windowing/WindowWordCount.java:
```diff
@@ -17,12 +17,13 @@
+import org.apache.flink.api.common.functions.ReduceFunction;
 import org.apache.flink.api.java.tuple.Tuple2;
-import org.apache.flink.api.java.utils.ParameterTool;
-import org.apache.flink.examples.java.wordcount.util.WordCountData;
```
well these changes don't belong here :)

aljoscha commented:
Jip, but that's not done very often, only when instantiating the `MetricRegistry`

zentol commented on flink-metrics/flink-metrics-statsd/src/test/java/org/apache/flink/metrics/statsd/StatsDReporterTest.java:
```diff
@@ -138,12 +138,11 @@ public void testStatsDHistogramReporting() throws Exception {
 
 			Configuration config = new Configuration();
 			config.setString(ConfigConstants.METRICS_REPORTERS_LIST, "test");
 			config.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test.class", StatsDReporter.class.getName());
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

zentol commented on flink-core/src/test/java/org/apache/flink/metrics/MetricRegistryTest.java:
```diff
@@ -42,7 +42,8 @@
 
-		config.setString(ConfigConstants.METRICS_REPORTER_CLASS, TestReporter1.class.getName());
+		config.setString(ConfigConstants.METRICS_REPORTERS_LIST, "test");
+		config.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test.class", TestReporter1.class.getName());
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

zentol commented on flink-core/src/test/java/org/apache/flink/metrics/MetricRegistryTest.java:
```diff
@@ -88,8 +91,10 @@ public void open(Configuration config) {
-		config.setString(ConfigConstants.METRICS_REPORTER_CLASS, TestReporter3.class.getName());
-		config.setString(ConfigConstants.METRICS_REPORTER_INTERVAL, "50 MILLISECONDS");
+		config.setString(ConfigConstants.METRICS_REPORTERS_LIST, "test");
+		config.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test.class", TestReporter3.class.getName());
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

zentol commented on flink-core/src/test/java/org/apache/flink/metrics/reporter/JMXReporterTest.java:
```diff
@@ -79,19 +81,24 @@ public void testGenerateName() {
-		cfg.setString(ConfigConstants.METRICS_REPORTER_CLASS, TestReporter.class.getName());
+		cfg.setString(ConfigConstants.METRICS_REPORTERS_LIST, "test1,test2");
+
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test1.class", JMXReporter.class.getName());
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

zentol commented on flink-core/src/test/java/org/apache/flink/metrics/reporter/JMXReporterTest.java:
```diff
@@ -79,19 +81,24 @@ public void testGenerateName() {
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test1.class", JMXReporter.class.getName());
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test1.port", "9020-9035");
+
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test2.class", JMXReporter.class.getName());
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

zentol commented on flink-core/src/test/java/org/apache/flink/metrics/reporter/JMXReporterTest.java:
```diff
@@ -128,18 +133,26 @@ public Integer getValue() {
+
+		cfg.setString(ConfigConstants.METRICS_REPORTERS_LIST, "test1,test2");
+
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test1.class", JMXReporter.class.getName());
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

zentol commented on flink-core/src/test/java/org/apache/flink/metrics/reporter/JMXReporterTest.java:
```diff
@@ -128,18 +133,26 @@ public Integer getValue() {
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test1.class", JMXReporter.class.getName());
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test1.port", "9040-9055");
+
+		cfg.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test2.class", JMXReporter.class.getName());
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

zentol commented on flink-core/src/test/java/org/apache/flink/metrics/reporter/JMXReporterTest.java:
```diff
@@ -197,7 +204,8 @@ public void testHistogramReporting() throws Exception {
 			Configuration config = new Configuration();
-			config.setString(ConfigConstants.METRICS_REPORTER_CLASS, "org.apache.flink.metrics.reporter.JMXReporter");
+			config.setString(ConfigConstants.METRICS_REPORTERS_LIST, "jmx_test");
+			config.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "jmx_test.class", "org.apache.flink.metrics.reporter.JMXReporter");
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

zentol commented on flink-metrics/flink-metrics-dropwizard/src/test/java/org/apache/flink/dropwizard/ScheduledDropwizardReporterTest.java:
```diff
@@ -67,7 +67,9 @@ public void testAddingMetrics() throws NoSuchFieldException, IllegalAccessExcept
 
-		configuration.setString(ConfigConstants.METRICS_REPORTER_CLASS, "org.apache.flink.dropwizard.ScheduledDropwizardReporterTest$TestingScheduledDropwizardReporter");
+		configuration.setString(ConfigConstants.METRICS_REPORTERS_LIST, "test");
+		configuration.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test.class", "org.apache.flink.dropwizard.ScheduledDropwizardReporterTest$TestingScheduledDropwizardReporter");
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

zentol commented on flink-metrics/flink-metrics-statsd/src/test/java/org/apache/flink/metrics/statsd/StatsDReporterTest.java:
```diff
@@ -137,9 +140,11 @@ public void testStatsDHistogramReporting() throws Exception {
-			config.setString(ConfigConstants.METRICS_REPORTER_INTERVAL, "1 SECONDS");
-			config.setString(ConfigConstants.METRICS_REPORTER_ARGUMENTS, "--host localhost --port " + port);
+			config.setString(ConfigConstants.METRICS_REPORTERS_LIST, "test");
+			config.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "test.class", StatsDReporter.class.getName());
```
using `.class` instead of `METRICS_REPORTER_CLASS_SUFFIX`

aljoscha commented:
I rebased this on top of #2226

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistry.java:
```diff
@@ -74,81 +76,97 @@ public MetricRegistry(Configuration config) {
+						if (this.executor == null) {
+							executor = Executors.newSingleThreadScheduledExecutor();
+						}
+						LOG.info("Periodically reporting metrics in intervals of {} {}", period, timeunit.name());
```
this message should include the reporter name.

aljoscha replied:
Fixing, I'll include the configured reporter name as well as the class.

zentol commented:
A simple test to verify that multiple configured reporters are properly instantiated/notified would be good, otherwise +1.

aljoscha commented:
There are tests in `JMXReporterTest` that instantiate multiple reporters but I'll also add a simple on to `MetricRegistryTest`.

aljoscha commented:
I addressed the comments.

zentol commented:
the test doesn't verify that new metrics are properly forwarded to all reporters :)

aljoscha commented:
Ahh, you're right, I'll fix that.

aljoscha commented:
I added a new test, is that what you had in mind?

zentol commented:
yes, thank you.

aljoscha commented:
Ok, now we first have to merge the other PR. 😃
